### PR TITLE
Add support for KDECAN

### PR DESCRIPTION
**Pull request by OXINARF**
KDE Direct has annouced a new CAN protocol and has paid development of it in ArduPilot, here it finally is.

It's marked WIP as this needs some changes (namely semaphores) and testing with current master.

### CODE REVIEW

**Comment by liang-tang** [comment]
details of KDECAN protocol
[KDECAN_Bus_Protocol_1.0.0.pdf](https://github.com/ArduPilot/ardupilot/files/2508771/KDECAN_Bus_Protocol_1.0.0.pdf)

**Comment by rmackay9** [comment]
I've added a dev call topic because I've got a similar project to add support for another manufacturer's CAN ESCs and I'd like this one to pass its review so we can base our work on this.

**Comment by rmackay9** [comment]
@OXINARF, could you rebase this on master maybe?

**Comment by rmackay9** [comment]
I think more comments in the AP_KDECAN.cpp file would make it easier to understand for future developers.  There are almost no comments in the .cpp file.

File: libraries/AP_KDECAN/AP_KDECAN.cpp [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+
+                for (uint8_t esc_num = sending_esc_num; esc_num < _esc_max_node_id; esc_num++) {
+
+                    if ((_esc_present_bitmask & (1 << esc_num)) == 0) {
+                        continue;
+                    }
+
+                    be16_t kde_pwm = htobe16(output_buffer[esc_num]);
+
+                    if (hal.util->safety_switch_state() == AP_HAL::Util::SAFETY_DISARMED) {
```
Comment (rmackay9) [inline libraries/AP_KDECAN/AP_KDECAN.cpp]: I wonder if it's necessary to have this safety switch check this far down in the code.  It seems to me that there are already checks at a much high level (like in the AP_Motors library) that will remove the need for this.

**Comment by OXINARF** [comment]
Hi @rmackay9,

Yes, as written in OP, this needs to be adapted to recent changes in master. I won't be touching this until, at least, next week though.

I can add more comments for sure.

Regarding safety switch check, can you point where that is checked in AP_Motors? I didn't find it and honestly it would be wrong at that level. The safety switch is supposed to be kind of a hardware cut off, so only at the output level it should check if it should send the output or not.

**Comment by hsteinhaus** [comment]
Just clicked on the linked protocol document above. The first pages contains a heap of license terms and conditions, including statements prohibiting to "... modify, adapt, translate or create any derivative works of the Software...". 

How can such a protocol be integrated into a piece of software, that is licensed under the GPLv3?

**Comment by khancyr** [comment]
@hsteinhaus the implementation on ArduPilot is open source, the implementation on ESC and other KDE software aren't open source nor is the protocol !

**Comment by hsteinhaus** [comment]
To me it sounds like these conditions have the be agreed to before being allowed to use/implement the protocol. 

```
The KDECAN Protocol Software, documentation, images, content, interfaces, fonts and any other data ... are licensed, not sold, to you by KDE Direct for use only under the terms of this License. 
```

But I am not a lawyer, though.

**Comment by OXINARF** [comment]
@hsteinhaus Some points:

- this work was paid by KDE, as written in OP
- the protocol wasn't provided to me under that (or any other restrictive) license
- a protocol isn't software; so my idea has been that "KDECAN Protocol Software" refers to the KDECAN protocol implementation in the ESC firmware
- I believe it's commonly accepted these days that a protocol can't be protected, you can legally sniff your network and discover the meaning of bits on your own

**Comment by rmackay9** [comment]
ping!  If we could just get it compiling that'd be great..

**Comment by OXINARF** [comment]
@rmackay9 Finally got it to build! Sorry for the delay. I haven't added comments yet, but I'll do that tomorrow. Any specific area you think comments are needed? I guess the main `loop` part is the one that needs it the most.

File: libraries/AP_BoardConfig/canbus_driver.cpp [hunk 2]
```diff
@@ -16,22 +16,37 @@
...
+  #if APM_BUILD_TYPE(APM_BUILD_ArduCopter) || APM_BUILD_TYPE(APM_BUILD_ArduPlane) || APM_BUILD_TYPE(APM_BUILD_ArduSub)
+    #include <AP_KDECAN/AP_KDECAN.h>
+  #endif
 
 // table of user settable CAN bus parameters
 const AP_Param::GroupInfo AP_BoardConfig_CAN::Driver::var_info[] = {
     // @Param: PROTOCOL
     // @DisplayName: Enable use of specific protocol over virtual driver
     // @Description: Enabling this option starts selected protocol that will use this virtual driver
+    // @Values{Copter,Plane,Sub}: 0:Disabled,1:UAVCAN,2:KDECAN
```
Comment (tridge) [inline libraries/AP_BoardConfig/canbus_driver.cpp]: why specific to those vehicle types?

**Comment by OXINARF** [comment]
@rmackay9 I've added some comments to the code now. Let me know if you think it needs more somewhere.

@tridge It is simply because it relies on two things: AP_Motors library for pre-arm checks and motorX functions in SRV_Channel library.

**Comment by rmackay9** [comment]
@OXINARF, great thanks.  I've started building the other company's can ESC support on top of this PR.

**Comment by OXINARF** [comment]
@rmackay9 That was really last minute, one second later and it would be merged 🙂 

> let's make this PR work for all vehicles and remove the comments like, "// To be replaced with macro saying if KDECAN library is included"

How would it work for all vehicles though? Not all vehicles have motors...

> thread should run with PRIORITY_MAIN+1 instead of PRIORITY_CAN

Why? This is why we have a PRIORITY_CAN... If the HALs don't set it correctly then we should correct the HAL.

> _rc_out_sem.take(1) could be replaced with WITH_SEMAPHORE

Sorry, but I'll reject that. I have a profound disagreement with using blocking semaphores everywhere. I've made sure to use them where it makes sense.

> do we really want the safety switch check? "if (hal.util->safety_switch_state() == AP_HAL::Util::SAFETY_DISARMED)"

You've mentioned that this is in motors library but I can't find that code and when asked previously you couldn't link it to me either. If it's not present there, then we absolutely want this.

> the servo range is using individual MIN/MAX but I think we should be using a hardcoded 1000 to 2000 range

As discussed, I'll change it to get the range from ESC scaling in the HAL. I'll ask to do that in another PR soon though, this has been waiting here for reviews for a long long time and I don't want it to bitrot.

**Comment by tridge** [comment]
> Sorry, but I'll reject that. I have a profound disagreement with using blocking semaphores everywhere. I've made sure to use them where it makes sense.

it most definitely should be a blocking semaphore in this case. Happy to discuss in detail on mumble, but basically by using a timeout you are disabling the priority inheritance code in ChibiOS and adding timing jitter

**Comment by OXINARF** [comment]
With @rmackay9's agreement this has been merged. We've discussed improvements and I've asked to let those for future improvements.

> it most definitely should be a blocking semaphore in this case. Happy to discuss in detail on mumble, but basically by using a timeout you are disabling the priority inheritance code in ChibiOS and adding timing jitter

@tridge Not sure when it can be used then... I can't do Mumble today at this hour. I'm happy to discuss changes in this area on another day.

**Comment by rmackay9** [comment]
@OXINARF and I discussed this a bit on skype and I *think* we agreed that for practical reasons at least we should change this KDECAN driver to use "rc output shadowing" to be consistent with how SBUS output and ToshibaCAN drivers work.  This means for example that ESC 1 will shadow the the same output sent SERVO1.

**Comment by rmackay9** [comment]
woohoo!  great to have this in!

**Pull request closed without merging**